controllers/appmesh/cloudmap_controller.go

```diff
@@ -0,0 +1,888 @@
+// +kubebuilder:rbac:groups=appmesh.k8s.aws,resources=virtualnodes/status,verbs=get
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;list;watch
+
+func (r *cloudMapReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	return runtime.HandleReconcileError(r.reconcile(req), r.log)
+}
+
+func (r *cloudMapReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appmesh.VirtualNode{}).
+		Watches(&source.Kind{Type: &corev1.Pod{}}, r.enqueueRequestsForPodEvents).
+		Complete(r)
+}
+
+func (r *cloudMapReconciler) reconcile(req ctrl.Request) error {
+	ctx := context.Background()
+	log := r.log.WithValues("Virtualnode", req.NamespacedName)
+
+	vNode := &appmesh.VirtualNode{}
+	if err := r.k8sClient.Get(ctx, req.NamespacedName, vNode); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	log.V(2).Info("VNode: ", "awsName: ", vNode.Spec.AWSName,
+		"vNode.PodSelector: ", vNode.Spec.PodSelector)
+	if !vNode.DeletionTimestamp.IsZero() {
+		return r.deleteCloudMapResources(ctx, vNode)
+	}
+	return r.reconcileVirtualNodeWithCloudMap(ctx, vNode)
+}
+
+func (r *cloudMapReconciler) reconcileVirtualNodeWithCloudMap(ctx context.Context, vNode *appmesh.VirtualNode) error {
+
+	if !r.isCloudMapEnabledForVirtualNode(vNode) {
+		return errors.Errorf("CloudMap not enabled for virtualNode: %s", vNode.Name)
+	}
+
+	if !r.isPodSelectorDefinedForVirtualNode(vNode) {
+		return errors.Errorf("PodSelector not defined for virtualNode: %s", vNode.Name)
+	}
+
+	cloudmapConfig := vNode.Spec.ServiceDiscovery.AWSCloudMap
+	creatorRequestID := vNode.UID
+
+	var podsList corev1.PodList
+	var listOptions client.ListOptions
+	listOptions.LabelSelector, _ = metav1.LabelSelectorAsSelector(vNode.Spec.PodSelector)
+	listOptions.Namespace = vNode.Namespace
+
+	if err := r.k8sClient.List(ctx, &podsList, &listOptions); err != nil {
+		r.log.Error(err, "Couldn't retrieve pods for VirtualNode")
+		return err
+	}
+
+	r.log.V(2).Info("Pods found matching this VNode: ", "Total Pods: ", len(podsList.Items))
+	if err := r.finalizerManager.AddFinalizers(ctx, vNode, k8s.FinalizerAWSCloudMapResources); err != nil {
+		r.log.Error(err, "..while adding cloudmap resource finalizer")
+		return err
+	}
+
+	_, err := r.createCloudMapService(ctx, cloudmapConfig, string(creatorRequestID))
+	if err != nil {
+		r.log.Error(err, "failed to create cloudMap Service for", "vNode: ", vNode.Name)
+		return err
+	}
+
+	serviceInstanceSummary := make(map[string]*cloudMapInstanceInfo)
+	for _, pod := range podsList.Items {
+		if pod.DeletionTimestamp != nil {
+			r.log.Info("vNode:", "Pod is being deleted: ", pod.Name)
+			continue
+		}
+
+		instanceId := r.podToInstanceID(&pod)
+		if instanceId == "" {
+			r.log.Info("No IP Address assigned to Pod:", pod.Name, "..Skipping for now")
+			continue
+		}
+
+		instanceInfo := &cloudMapInstanceInfo{}
+		if r.isPodReady(&pod) {
```

Comment:
better to separate a component to get backend pods for a virtualNodes.
some thing like ```
type virtualNodeEndpointResolver interface {
     Resolve(ctxcontext.context, vn *appmesh.VirtualNode) (ready []*corev1.Pod, notReady[]corev1.Pod, ignoredPods []corev1.Pod)
}
```

Reply:
done